db/db_iter.cc

```diff
@@ -470,11 +483,19 @@ inline bool DBIter::FindNextUserEntryInternal(bool skipping, bool prefix_check)
     }
 
     if (IsVisible(ikey_.sequence)) {
-      if (skipping && user_comparator_.Compare(ikey_.user_key,
-                                               saved_key_.GetUserKey()) <= 0) {
+      // If the previous entry is of seqnum 0, the current entry will not
+      // possibly be skipped. This condition can potentially be relaxed to
+      // prev_key.seq <= ikey_.sequence. We are cautious because it will be more
+      // prone to bugs causing the same user key with the same sequence number.
+      if (!is_prev_key_seqnum_zero && skipping &&
+          user_comparator_.Compare(ikey_.user_key, saved_key_.GetUserKey()) <=
```

Comment:
It is not clear to me why we need a global is_key_seqnum_zero_ variable? Can't we simply conpare ikey_.sequence with saved_key_.sequence?

Reply:
Unfortunately do not always have `saved_key_.sequence`. We sometimes to `savec_key_.SaveUserKey()` in which case the internal part of the key is not available. Correct me if I'm wrong.

Reply:
You are right! But do you know why? Would not saving the sequence make much difference in performance?

Reply:
I look at the code a little bit. Changing the code to make `saved_key_` is always internal key doesn't seem to be trivial. `ikey_` is a `ParsedInternalKey` where user key and internal bytes are not stored together. And sometimes we don't copy the key into `saved_key_` but provide a pointer to it. Since `ikey_` doesn't provide a counter to internal key, we need to find internal keys from where `saved_key_` got key from. I don't know whether we'll hit other problem, but even if there isn't a problem, the code change is going to be very dramatic. I don't think we should take a risk doing anything like it.

Reply:
> And sometimes we don't copy the key into saved_key_ but provide a pointer to it. 

I believe the same could be done for the internal key as well.

Agreed. Was not suggesting to do that here. Wanted to understand the code better for future refactoring.